### PR TITLE
Fixes and enhancements in JSON Configuration schema generator

### PR DESCRIPTION
**Pull request by bart-vmware**
This PR contains bugfixes and enhancements to the JSON Configuration schema generator, which I'm trying to use outside of Aspire. I've tested with VS 2022 v17.10.1. Please see the individual commit messages for change details. Hopefully, this PR brings the tool a step closer to achieving #3309.

I'm curious as to why JSON properties are sorted alphabetically. I found it makes it harder to compare against the original source code it was generated from.

> [!NOTE]
> This includes a fix in the RuntimeSource directory to prevent a `StackOverflowException`. It should probably be synced with the original runtime source code. Tracked at https://github.com/dotnet/runtime/issues/103802.

/cc @eerhardt 
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4441)

### CODE REVIEW

**Comment by eerhardt** [comment]
> I'm curious as to why JSON properties are sorted alphabetically.

It makes it "stable", so the .json file can be checked in. It makes version-to-version updates of a library easier, so when new properties are added, you can see what properties are added, and they don't become all mixed up. See https://github.com/dotnet/aspire/pull/2663/ for an example, a new property was added and we can easily diff and see the new property (the properties aren't in a new order after the update).

File: src/Components/Aspire.Azure.Storage.Blobs/ConfigurationSchema.json
Comment (eerhardt) [inline src/Components/Aspire.Azure.Storage.Blobs/ConfigurationSchema.json:78]: We left these newlines in purposefully so when they are in the docs, they are in the UI as well.

Today:
image

After this change:

image

This makes it less readable.

cc @sebastienros

File: src/Tools/ConfigurationSchemaGenerator/ConfigurationSchemaGenerator.csproj
Comment (eerhardt) [inline src/Tools/ConfigurationSchemaGenerator/ConfigurationSchemaGenerator.csproj:23]: I don't think we should be using this. We should only be using public APIs.

Which internal APIs are you using here?

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaGenerator.cs
Comment (eerhardt) [inline src/Tools/ConfigurationSchemaGenerator/ConfigSchemaGenerator.cs:71]: I don't think path should ever be null or empty. What's the scenario here?

File: tests/ConfigurationSchemaGenerator.Tests/RecordWithPrimitives.cs
Comment (eerhardt) [inline tests/ConfigurationSchemaGenerator.Tests/RecordWithPrimitives.cs:12]: Can you leave the existing tests in place? You can add new ones, but removing existing tests makes it hard to see what changed.

We are losing some coverage here.

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs
Comment (sebastienros) [inline src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs:84]: Does the immutability matter?
I couldn't find any place where it would.

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaGenerator.cs
Comment (bart-vmware) [reply src/Tools/ConfigurationSchemaGenerator/ConfigSchemaGenerator.cs:71]: This is new: an empty path allows to define properties in the configuration root. See [ShouldAllowEmptyComponentPath](https://github.com/dotnet/aspire/blob/bce9dd1249c73123eb260362ea873d822ebf2493/tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs#L719-L754).

File: src/Tools/ConfigurationSchemaGenerator/ConfigurationSchemaGenerator.csproj
Comment (bart-vmware) [reply src/Tools/ConfigurationSchemaGenerator/ConfigurationSchemaGenerator.csproj:23]: `Microsoft.CodeAnalysis.Shared.Extensions.ISymbolExtensions.GetDocumentationComment`, this is explained in https://github.com/dotnet/aspire/pull/4441/commits/218d43de70c75637b767198d6211f1e11828466f.

File: tests/ConfigurationSchemaGenerator.Tests/RecordWithPrimitives.cs
Comment (bart-vmware) [reply tests/ConfigurationSchemaGenerator.Tests/RecordWithPrimitives.cs:12]: I'm convinced that no coverage is lost. I moved the existing tests for primitive types from `RecordWithPrimitiveTypes` to [ShouldConvertTypes](https://github.com/dotnet/aspire/blob/bce9dd1249c73123eb260362ea873d822ebf2493/tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs#L158-L418). The [diff](https://github.com/dotnet/aspire/compare/ace04ef35b8c08e302e6c6dc34a486c6c76447a5...bart-vmware:aspire:config-schema-types#diff-5231539b68734fdabae038ece33d44afd5377e7810b7fd9962382a31b9d53306) between the first and last commit shows how this PR changes them. The only change in output is that `object?` gets converted to `type:object`, which used to be `type:string`.

The reason I moved these is because I had to insert more types. Then the numbering in property names got in the way, so I kept renumbering them, which messes up the PR diff. At that point, I realized it makes more sense to have unit tests to cover all the primitive types, than to stash everything into one very large integration test.

But my move did not cover associated doc-comments. So I picked a few basic types with interesting JSON structures and included doc-comments for them in the new integration test.

Another way of looking at the diff is [here](https://github.com/dotnet/aspire/compare/main...bart-vmware:aspire:config-schema-types-old-tests#diff-ba37d348ecc89e724d88d88a3e2dc1f78c95c76c456d1a04bd359f1a7c3f2a9c), where I added an extra commit on top of this PR that restores the original `RecordWithPrimitives.cs` file. It shows the same change as mentioned above.

This PR greatly improves coverage, and I think the new structure makes it easier to keep it high. Please let me know if you still want me to re-add `RecordWithPrimitives.cs`.

File: src/Components/Aspire.Azure.Storage.Blobs/ConfigurationSchema.json
Comment (bart-vmware) [reply src/Components/Aspire.Azure.Storage.Blobs/ConfigurationSchema.json:78]: That's interesting. The _duplicate_ line break at https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/src/BlobClientOptions.cs#L171-L180 shows up at https://learn.microsoft.com/en-us/dotnet/api/azure.storage.blobs.blobclientoptions.georedundantsecondaryuri?view=azure-dotnet#azure-storage-blobs-blobclientoptions-georedundantsecondaryuri, but the _singular_ line breaks do not. Are you okay with preserving only the duplicate line breaks?

**Comment by bart-vmware** [comment]
> > I'm curious as to why JSON properties are sorted alphabetically.
> 
> It makes it "stable", so the .json file can be checked in. It makes version-to-version updates of a library easier, so when new properties are added, you can see what properties are added, and they don't become all mixed up. See #2663 for an example, a new property was added and we can easily diff and see the new property (the properties aren't in a new order after the update).

Thanks for the clarification. I suppose it depends on whether to optimize for the development of the generator versus its consumption. Although member order in assemblies is already deterministic, according to the Remarks section at https://learn.microsoft.com/en-us/dotnet/api/system.type.getproperties?view=net-8.0#system-type-getproperties:

> In .NET 6 and earlier versions, the [GetProperties](https://learn.microsoft.com/en-us/dotnet/api/system.type.getproperties?view=net-8.0) method does not return properties in a particular order, such as alphabetical or declaration order. Your code must not depend on the order in which properties are returned, because that order varies. However, starting with .NET 7, the ordering is deterministic based upon the metadata ordering in the assembly.

Because https://github.com/dotnet/aspire/pull/2663/files#diff-ba29f362bde4fa8dfaf456726c5d33b10aec019ce5485384d75974e2c98b9277 did not shuffle the order of existing properties, the diff in Aspire would still remain minimal.

**Comment by bart-vmware** [comment]
Regarding the failed checks: I've followed the steps at https://github.com/dotnet/aspire/tree/main/tests/Aspire.Workload.Tests#readme and can't reproduce the failure locally. Here's what's in the failure log:
```
   Failed Aspire.Workload.Tests.StarterTemplateWithWithRedisCacheTests.WebFrontendWorks(urlPrefix: "https://") [13 s]
  Error Message:
   Microsoft.Playwright.PlaywrightException : net::ERR_NETWORK_CHANGED at https://localhost:17003/login?t=d2e8dbf1418a5877c9cd259dff91a759
```
I don't suspect this is related to the changes in this PR, but please let me know if there's anything I can do to get a green build.

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs
Comment (eerhardt) [inline src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs:433]: This algorithm is a bit surprising. We inject `<br/><br/>` in ConvertBlankLines and then here turn them to `\n\n\n\n`, and then `\n\n`.

Can you add some comments here to explain it?

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs
Comment (eerhardt) [inline src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs:386]: It looks like we lost this TODO and handling for `EnumerableSpec`. Can we add it back?

**Review by eerhardt — COMMENTED** [review]
Sorry for the delay here @bart-vmware, I was on vacation for a while. But I'm back now.

First off, this is really great! Thank you for the contribution here. These fixes and enhancements really help.

I pushed some changes that I could (mostly to get rid of the extra dependency and to just use normal reflection). I left some other comments. Once those are addressed, I can merge this.

Thanks again.

File: tests/ConfigurationSchemaGenerator.Tests/RecordWithPrimitives.cs
Comment (eerhardt) [reply tests/ConfigurationSchemaGenerator.Tests/RecordWithPrimitives.cs:12]: Thanks for the explanation. I agree this coverage isn't lost (and is greatly expanded in the new code).

File: src/Tools/ConfigurationSchemaGenerator/ConfigurationSchemaGenerator.csproj
Comment (eerhardt) [reply src/Tools/ConfigurationSchemaGenerator/ConfigurationSchemaGenerator.csproj]: I removed this and just used regular reflection.

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs
Comment (bart-vmware) [reply src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs:386]: There's no need for that. The behavior is the same for its base type, `CollectionSpec`:

image

The test coverage is at `GeneratorTests.ShouldConvertTypes`:
```c#
[InlineData("IEnumerable<string>?",
    """
    "type": "array",
    "items": {
      "type": "string"
    }
    """)]
```

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs
Comment (bart-vmware) [reply src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs:100]: In:
```c#
[assembly: ConfigurationSchema("One.Two.Three", typeof(ExampleSettings))]
```
"One", "Two" and "Three" are components.

File: tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs
Comment (bart-vmware) [reply tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs:491]: In general, get-only properties are excluded in the schema. But collections and non-primitive types are exceptions to that. This test verifies the property gets emitted, even if the declared type is get-only.

I didn't assert on the full tree, because most of its content isn't relevant here. This way, future refactorings of other tree parts won't require "fixing" this test.

File: src/Tools/ConfigurationSchemaGenerator/RuntimeSource/Configuration.Binder/ConfigurationBindingGenerator.Parser.cs
Comment (bart-vmware) [reply src/Tools/ConfigurationSchemaGenerator/RuntimeSource/Configuration.Binder/ConfigurationBindingGenerator.Parser.cs:531]: Created https://github.com/dotnet/runtime/issues/103802.

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs
Comment (bart-vmware) [reply src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs:116]: I've pushed a commit that explains it.

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs
Comment (bart-vmware) [reply src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs]: Good catch! I took this from the original regex without thinking about it. Pushed a commit to address this.

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs
Comment (bart-vmware) [reply src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs:440]: I've pushed a commit to explain. Tests exist for all whitespace replacements around `<br/>`.

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs
Comment (eerhardt) [reply src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs:386]: Sounds good. Thanks for the explanation.

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs
Comment (eerhardt) [reply src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs:100]: Can we use a different term? "Components" is kind of an overloaded term in the .NET Aspire world. How about "GeneratePathSegment"?

File: tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs
Comment (eerhardt) [reply tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs:491]: But it doesn't catch other bugs. It also doesn't validate that "FormatSettings" gets put in the right location. Since we are validating the full tree everywhere else, it feels like we should just do the same here. It gets us better test coverage.

File: tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs
Comment (eerhardt) [reply tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs:528]: My biggest concern is that this comment doesn't match the code below. I don't see it using `TestProperty1` or `TestProperty2`. I think the comment is confusing since it doesn't match.

File: src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs
Comment (eerhardt) [reply src/Tools/ConfigurationSchemaGenerator/ConfigSchemaEmitter.cs:116]: Thanks!

File: src/Tools/ConfigurationSchemaGenerator/RuntimeSource/Configuration.Binder/ConfigurationBindingGenerator.Parser.cs
Comment (eerhardt) [reply src/Tools/ConfigurationSchemaGenerator/RuntimeSource/Configuration.Binder/ConfigurationBindingGenerator.Parser.cs:531]: Awesome. Thank you!

File: tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs
Comment (eerhardt) [reply tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs:1222]: > For example, leaving out <PreserveCompilationContext>true</PreserveCompilationContext> in the test project fails for reasons I don't fully understand.

Check out https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#preservecompilationcontext for an explanation of what PreserveCompilationContext.

We use this in the tests so we can get all the "ref" assemblies that were used at build-time. We can load the "ref" assemblies up when the test runs and pass them to Roslyn, so Roslyn can return valid information.

File: tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs
Comment (bart-vmware) [reply tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs]: Makes sense. I've replaced the comment with compilable code.

File: tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs
Comment (bart-vmware) [reply tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs]: Existing tests already cover all other concerns. It's _duplicating_ test coverage, not making it better. But I've added it anyway.

**Comment by bart-vmware** [comment]
@eerhardt I've addressed all comments. Can you take another look? Please don't hesitate to resolve the open conversations if they've been addressed properly.

I don't think the failing checks are related to these changes, but I can't rerun them.

File: tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs
Comment (eerhardt) [reply tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs:1334]: > Do you understand why that happens?

yes, this happens because not all the "ref" assemblies are available when you remove PreserveCompilationContext.

> The integration test doesn't load reference assemblies from the "ref" directory anymore

This isn't true because the code recurses to all directories under the current directory, which includes the `ref` directory.

**Review by eerhardt — APPROVED** [review]
Thanks again for this great contribution, @bart-vmware!

The changes look good to me. I'm ready to merge once CI is green. We had some flaky failures in the past, but we should be in a better state now.

File: tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs
Comment (bart-vmware) [reply tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs:1334]: Thanks for the explanation, makes sense.

**Pull request merged**
